Closes #16 Remove diagonallayout

marioguillermo opened the pull request:
Closes #16 Remove diagonallayout 
Replace diagonalLayout library with a custom view implementation

Commit:
Closes #16 Remove diagonallayout

Commit:
Remove repeated assignation of style

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+
+    private lateinit var arcPaint: Paint
+    private lateinit var path1: Path
+    private lateinit var path2: Path
```
Podemos agregar una línea vacía entre las propiedades privadas y esta variable?

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+    private lateinit var path1: Path
+    private lateinit var path2: Path
+    @ColorInt
+    var triangleColor = Color.BLACK
```
Creo que podemos ahorrarnos la función `changeTriangleColor` si trabajamos con el `set` de la variable, algo como:

```
private var title: String? = null
        set(value) {
            field = value
            componentTitle.text = title
        }
```

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+        setup(attrs)
+    }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
```
Sobre los constructores también podemos ahorrarnos las multiples implementaciones usando algo así:

```
class MyComponent @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        layoutId: Int,
        defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {
```

No sé si para el  `View` puede aplicar algo similar 🤔 

Me parece que este enfoque también obliga a tener un método `init`

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private fun setup(attrs: AttributeSet?) {
+        val typedArray: TypedArray? = context.obtainStyledAttributes(attrs, R.styleable.TriangularView)
```
¿Qué opinas de agregar estas extensiones?

```
/**
 * Read XML attributes for a custom component.
 *
 * Usually on custom widget components we should have and "attrs" variable passed to the
 * class constructor. Given a custom "styleable" you can call this method to retrieve the
 * XML attributes as a [TypedArray]
 */
fun AttributeSet.asTypeArray(context: Context, styleableResId: IntArray): TypedArray? {
    return context.theme.obtainStyledAttributes(
            this, styleableResId, 0, 0
    )
}

/**
 * Allows to access [TypedArray] methods in order to retrieve data by wrapping the
 * [AttributeSet.asTypeArray] method.
 */
inline fun AttributeSet.consumeTypeArray(
        context: Context,
        styleableResId: IntArray,
        consume: TypedArray.() -> Unit = {}) {

    asTypeArray(context, styleableResId)?.apply {
        consume()
        recycle()
    }
}
```

Yo las tengo en un file `ComponentExtensions` pero puede llevar otro nombre, su uso sería así:

```
 init {
        inflate(layoutId, context)
        attrs?.consumeTypeArray(context, R.styleable.MyStyleable) {
            retrieveAttributes(this)
        }
    }
```

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+        typedArray?.getInt(R.styleable.TriangularView_triangleColor, 0)?.let { triangleColor = it }
+        typedArray?.recycle()
+        path1 = Path()
+        path2 = Path()
```
Me parece que el `path2` no es utilizado en la clase 🤔 ¿Podríamos eliminarlo y renombrar `path1` por `path` ?

luisburgos commented on app/src/main/java/mx/yellowme/youst/common/components/TriangularView.kt:
```diff
@@ -0,0 +1,77 @@
+        setBackgroundColor(Color.TRANSPARENT)
+        val canvasWidth = canvas.width.toFloat()
+        val canvasHeight = canvas.height.toFloat()
+        arcPaint.apply {
```
Usar `.run` en lugar de apply podría funcionar aquí? Lo comento por que me parece que apply nos funciona mejor cuando usamos la instancia que nos regresa para almacenarla 🤔 

Mismo comentario para el  `apply` usado en `path1`

Commit:
Replace apply with run

Commit:
Improve TriangularView component

Add extension to read attributes for the view
Add Overload constructor to the TriangularView
Remove path variable
Refactor path1 to path

Commit:
Add comment regions to TriangularView
